### PR TITLE
Add RUS to balance cross validation for classification problems

### LINKED ISSUE #973: Balanced sampling for classification problems
#971 tracks adding a data check to detect and warn about severe class-imbalance. #457 tracks adding an automl objective which can perform well on imbalanced datasets

Detecting and warning about class imbalance could be helpful, but ideally we could improve our classifier quality by down-sampling the training data to flatten the distribution of the target over each class.

There's a couple levels this problem exists at:
1. In automl, if we want to make a fair comparison between models, we should either use an automl objective which is hardened to imbalanced data (like Neyman-Pearson tracked by #457), or we should balance the data before computing the objective score.
2. During the training of each pipeline, each CV fold should be down-sampled to have a balanced distribution of labels. A simple way to do this is to compute the splits in automl as we do now, apply down-sampling to balance, and then train/evaluate each pipeline on each split. We could also have the pipelines generate the splits and down-sample internally, but if this is done right there wouldn't be any difference in accuracy.

I see [sklearn's doc mentions class imbalance](https://scikit-learn.org/stable/modules/cross_validation.html#cross-validation-iterators-with-stratification-based-on-class-labels), but I don't agree with their recommendation "to ensure that relative class frequencies is approximately preserved in each train and validation fold." I suppose we could put this to the test and see which strategy seems to generalize more effectively.

### PR DESCRIPTION
**Pull request by jeremyliweishih**
Fixes #973.

### CODE REVIEW

File: evalml/automl/automl_search.py [hunk 1]
```diff
@@ -612,13 +613,21 @@ def _compute_cv_scores(self, pipeline, X, y):
             logger.debug(f"\t\tTraining and scoring on fold {i}")
             X_train, X_test = X.iloc[train], X.iloc[test]
             y_train, y_test = y.iloc[train], y.iloc[test]
+
             if self.problem_type in [ProblemTypes.BINARY, ProblemTypes.MULTICLASS]:
+                try:
+                    rus = RandomUnderSampler(random_state=self.random_state, replacement=False)
+                    X_train, y_train = rus.fit_resample(X_train, y_train)
+                except Exception:
```
Comment (gsheni) [inline evalml/automl/automl_search.py:621]: Is it possible to catch a specific Exception, rather than blanket exception?

File: evalml/automl/automl_search.py [hunk 2]
```diff
@@ -612,13 +613,21 @@ def _compute_cv_scores(self, pipeline, X, y):
             logger.debug(f"\t\tTraining and scoring on fold {i}")
             X_train, X_test = X.iloc[train], X.iloc[test]
             y_train, y_test = y.iloc[train], y.iloc[test]
+
             if self.problem_type in [ProblemTypes.BINARY, ProblemTypes.MULTICLASS]:
+                try:
+                    rus = RandomUnderSampler(sampling_strategy='not minority' ,random_state=self.random_state, replacement=False)
```
Comment (jeremyliweishih) [inline evalml/automl/automl_search.py]: https://imbalanced-learn.org/stable/under_sampling.html#prototype-selection

File: core-requirements.txt [hunk 3]
```diff
@@ -10,3 +10,4 @@ psutil>=5.6.3
 requirements-parser>=0.2.0
 shap>=0.35.0
 texttable>=1.6.2
+imbalanced-learn>=0.7.0
```
Comment (dsherry) [inline core-requirements.txt:13]: @jeremyliweishih what is this?

File: evalml/automl/automl_search.py [hunk 4]
```diff
@@ -612,13 +613,21 @@ def _compute_cv_scores(self, pipeline, X, y):
             logger.debug(f"\t\tTraining and scoring on fold {i}")
             X_train, X_test = X.iloc[train], X.iloc[test]
             y_train, y_test = y.iloc[train], y.iloc[test]
+
             if self.problem_type in [ProblemTypes.BINARY, ProblemTypes.MULTICLASS]:
+                try:
+                    rus = RandomUnderSampler(sampling_strategy='not minority', random_state=self.random_state, replacement=False)
+                    X_train, y_train = rus.fit_resample(X_train, y_train)
+                except ValueError:
+                    logger.warning(f"\t\t\tFold {i}: could not downsample to balance classification")
```
Comment (dsherry) [inline evalml/automl/automl_search.py:622]: Under what circumstances would this error be raised?

Comment (jeremyliweishih) [reply evalml/automl/automl_search.py:622]: I ran into it when there wasn't enough samples for each class. Maybe I can put a datacheck before this but I only went for a warning for now.

File: core-requirements.txt (hunk 3 above)
Comment (jeremyliweishih) [reply core-requirements.txt:13]: The package that I'm using to down sample and balance classes! Using selection under sampling to do so: https://imbalanced-learn.readthedocs.io/en/stable/under_sampling.html#prototype-selection

**Comment by dsherry** [comment]
@jeremyliweishih given that we discussed this yesterday and have more perf testing to do, I pushed this back to draft.

File: evalml/automl/automl_search.py (hunk 4 above)
Comment (jeremyliweishih) [reply evalml/automl/automl_search.py:622]: @bchen1116 @dsherry @rpeck 

It seems imblearn does give an error if there are not enough samples. I think there are parameters to manage the number of samples as well.

**Comment by dsherry** [comment]
Closing in favor of other PRs for #973
